[그리디] 이창희 Spring Data Jpa 4,5,6 단계 제출합니다. (JPA Repository)

chxghee opened the pull request:
안녕하세요 경환님 😄
스터디를 시작한지 4개월 가량이 지났는데,
리뷰로 만나 뵙기는 처음이네요!! 

이번 미션 잘 부탁 드립니다~~

이번 미션은 기존 엔티티 매니저를 이용해 구현했던  4,5,6 단계 미션을 JpaRepository 로 바꾸어 보는 미션이었어요

JpaRepository는 메서드 이름으로 쿼리를 자동 생성해 주는 것 뿐만 아니라,
직접 엔티티 매니저 통해 구현할 때 사용했던 페치 조인, 락, 영속성 flush와 clear 등등..
어노테이션 기반으로 다 설정할 수 있게 구현되어 있다 보니
정말 편리하게 코드를 작성할 수 있게 해주는구나 하고 체감이 되었습니다 ㅎㅎ

아무래도 저번에 구현했던 코드를 JpaRepository 바꾸는 미션이었기 때문에 비교적 빨리 끝났던 것 같네요

제 코드에서 개선할 점이나 의문이 드는 부분이 있으시다면 코멘트 남겨 주시면 감사하겠습니다!!

P.S. 단계별 요구사항 구현 흐름은 리드미에 남겨 두었어요~

chxghee committed:
refactor: TimeRepository Spring Data Jpa 인터페이스로 전환

chxghee committed:
refactor: ThemeRepository Spring Data Jpa 인터페이스로 전환

chxghee committed:
refactor: MemberRepository Spring Data Jpa 인터페이스로 전환

chxghee committed:
refactor: ReservationRepository Spring Data Jpa 인터페이스로 전환

chxghee committed:
WaitingRepository, WaitingOrderCounterRepository Spring Data Jpa 인터페이스로 전환

chxghee committed:
refactor: 불필요한 임포트문 삭제 및 인터셉터 수정

chxghee committed:
docs: 리드미 수정

chxghee committed:
docs: 리드미 수정 2

dooboocookie commented on src/main/java/roomescape/waiting/domain/repository/WaitingOrderCounterRepository.java:
```diff
@@ -0,0 +1,23 @@
+@Repository
+public interface WaitingOrderCounterRepository extends JpaRepository<WaitingOrderCounter, Long> {
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
+    @Query("select c FROM WaitingOrderCounter c " +
+            "where c.themeId = :themeId " +
+            "and c.date = :date " +
+            "and c.timeId = :timeId")
+    Optional<WaitingOrderCounter> findByThemeIdAndDateAndTimeIdWithLock(Long themeId, String date, Long timeId);
```
## 👀 Comment

(만약 미션이 아닌 상황이면 Request Change에 가까웠을 것 같은데, 이 미션을 진행하면서 동시성에 대해 고민했다는 것 자체로 충분하다고 생각하기 때문에 Comment로 달겠습니다.)

#### 1️⃣ 왜 비관락?
동시성을 제어하는 방법은 여러가지 방법이 존재하고, 구글이나 지피티에 조금만 찾아봐도 여러 방식을 제안합니다.
아마 창희님도 여러 방법 중에 비관락을 선택하셨을 것 같습니다.
여러 방법 중에 왜 이 Waiting을 저장하는 로직이 비관락이 적합하다고 생각했는지 궁금해요.

#### 2️⃣ 동시 요청을하면, 나중 요청이 진짜 재시도를 하다 다음 대기번호를 가져가나?

- WaitingService.java
```java
    public WaitingResponse save(WaitingCommand command) {
        // ...
        // 카운터 생성 충돌시 재시도 로직
        while (attempts < maxAttempts) {
            attempts++;
            try {
                // ...
            } catch (DataIntegrityViolationException e) {
                if (attempts >= maxAttempts) throw new ApplicationException(WaitingException.RETRY_WAITING_ORDER_COUNTER_FAILED);
            }
        }
        throw new ApplicationException(WaitingException.RETRY_WAITING_ORDER_COUNTER_FAILED);
    }
```
현재 구조는 `Transactional(readonly=true) save() 메서드`가이고  비관락 에러를 발생할 수 있는 `Transactional create...WithLock() 메서드`를 호출하는데요.

제가 이를 테스트 해보진 않았지만, 제 지식 안에서 의도대로 동작하지 않을 것 같아서 몇가지 질문 드릴게요.

1. 왜 save 메서드는 리드온리인가?
2. [셀프 인보케이션](https://jas-note.tistory.com/111)에 해당되는가?
3. 셀프 인보케이션이 맞다면 의도한대로 동작하는가?
4. 이 동작들이 잘 실행되는지 테스트 짤 순 없을지?

질문에 답해보면서 아래 문제들을 잘 
- 이 로직이 잘 동작하는지 확인할 수 있는지?
- 이 작업이 얼마나 필요하고 가치가 있는지?
- 더 나은 해결책이 있는지?
 

#### 3️⃣ 비관락은 왜 예외를 던지는가?

일련의 과정에서 예외가 발생시키는 것은 매우 큰 일입니다.
예외가 발생하면, 
그대로 로직이 끝나고 예외 응답을 주기도 하고,
Transactinal 어노테이션은 예외를 catch 한다한들 rollback 플래그가 true로 바뀌기도 합니다.
그만큼 예외라함은 정상 상황 안에서 다른 분기라기 보단 비정상적인 상황이라고 생각하는 것이 적절하다고 봅니다.

비관락은 애초에 충돌이 발생할 가능성이 매우 크다는 가능성을 두고 예외를 던져버리는 기법입니다.
재시도를 할 수 있는 여러 방법은 존재하지만, 예외라는 큰 위험부담을 감수하고 이 예약의 순번을 부여해야하는지 잘 모르겠어요.

동시성을 해결하는 문제가 백엔드 개발자가 되기 위해 꼭 겪어봐야할 경험처럼 여겨지곤하는데요. (사실 저도 중요하다고 생각하긴 합니다.)
그래도 가장 중요한 것은 동시에 접근해도 문제 없는 API를 제공하고 동시성 문제가 생기지 않게 데이터를 관리하는 것이겠죠.
제가 봤을 때 이 예약 시스템은 충분히 비관락을 사용하지 않아도 Waiting 리스트를 잘 만들 수 있을거라 생각해요.

---
가장 눈에 띄는 부분이여서 여러 질문 드렸습니다.
코멘트 읽어보시면서 조금 더 고민해보시면 좋겠습니다.

chxghee replied:
## 1️⃣ 왜 비관락?
> 동시성을 제어하는 방법은 여러가지 방법이 존재하고, 구글이나 지피티에 조금만 찾아봐도 여러 방식을 제안합니다.
아마 창희님도 여러 방법 중에 비관락을 선택하셨을 것 같습니다.
여러 방법 중에 왜 이 Waiting을 저장하는 로직이 비관락이 적합하다고 생각했는지 궁금해요.

우선 예약 대기 요구사항을 처음 구현한 방식은
`Waiting`테이블에 예악 순서를 저장하지 않고, **조회 시** Id 기준으로 하여 그때 그때 랭킹을 계산하는 방식을 사용했어요

떄문에 별도로 락을 걸지 않아도 동시성 이슈가 발생하지 않았는데요, 하지만 결론적으로 락을 거는 방식으로 변경을 했어요.

그러한 선택을 한 이유는 다음과 같아요.

```sql
select new WaitingWithRank(w,(
                      select count(w2) + 1 
                      from Waiting w2 
                      where w2.theme = w.theme 
                              and w2.date = w.date 
                              and w2.time = w.time 
                              and w2.id < w.id
                     )
        )
from Waiting w 
where w.member.id = :memberId
```
조회 시에 랭킹을 계산하기 위해서 위의 jpql을 사용했었는데,
이 방식은 내 예약 대기 리스트만큼 서브쿼리가 실행되어,  먼저 요청되었던 예약 대기 개수를 세는 구조였습니다.
그렇기 떄문에 사용자가 예약 대기를 많이 했다면, 조회시 성능이 많이 나쁠것 같다라는 결론을 내리게 되었어요

아무래도 조회되는 횟수가 등록하는 횟수보다 훨씬 많을 거기 때문에 조회 성능을 올리고 싶었고,
그래서 순위를 유지하는 `waiting_order` 컬럼을 추가하여 문제를 해결하는 방향으로 전환했어요.

하지만 이렇게 순위 상태를 컬럼으로 유지하게 되면 동시성 문제가 발생하게 되었어요.

어떤 방식을 사용해서 동시성 문제를 해결할까 고민했었는데,
예약 대기의 요구사항의 특성이 다음과 같다고 생각했어요.
1. 순위 중복이 없어야 한다.
2. 동시 예약이 발생 하더라도 요청이 실패하면 안된다.
3. 반드시 한 명씩 순차적으로 저장되고 DB에 반영되어야 한다.

이렇게 생각 했을때 가장 잘 들어 맞는 방식이 `비관적 락`을 거는 것이라고 판단해서 적용해 보앗습니닷

chxghee replied:
## 2️⃣ 동시 요청을하면, 나중 요청이 진짜 재시도를 하다 다음 대기번호를 가져가나?
> 1. 왜 save 메서드는 리드온리인가?
> 2. [셀프 인보케이션](https://jas-note.tistory.com/111)에 해당되는가?
> 3. 셀프 인보케이션이 맞다면 의도한대로 동작하는가?
> 4. 이 동작들이 잘 실행되는지 테스트 짤 순 없을지?


수요일에 이 미션을 제출 하고, 제 코드를 한번 돌아봤는데요,
저도 이 부분에 있어 뭔가 잘못되었다...! 하는 생각이 들었어요.

<img width="2220" height="1574" alt="image" src="https://github.com/user-attachments/assets/0143d90c-a72f-4173-ac78-460633dce8fe" />

대충 이런 상황인데요,

우선 이렇게 코드를 작성했던 의도는 
조회는 트랜잭션에서 제외를 하고, 트랜잭션 범위를 줄여 커녁선을 점유하는 시간을 줄이고자 했어요.

---

우선 여기서 첫번째 잘못된 점이 있었어요
1. 클래스 레벨에 `@Transactional(readonly=true)` 옵션이 붙어 `save()`메서드에 읽기 전용 트랜잭션이 시작되었다.

원래 의도는 조회 로직은 트랜잭션에서 제외하는 것이였는데, 클래스레벨에 붙은 어노테이션 때문에 트랜잭션이 시작되어 버렸어요

<br>
<br>

---
그리고 두 번째로는 다음의 문제가 있었어요.

2. `save()`메서드의 내부에서, 같은 클래스 에 위치한 `트랜잭션이 붙은 메서드`(create...WithLock())를 호출하게 되어 프록시 객체로 트랜잭션을 동작 시킬 수 없었다.

즉, 경환님께서 언급해 주신 `셀프 인보케이션`문제가 발생하게 되었어요
제 의도가 제대로 동작하지 않는 좀더 근본적인 이유에 해당하는 것 같아요.


JPA 가 트랜잭션 어노테이션을 처리하기 위해 AOP를 기반으로 동작하게 되는데,
[@Transactional 동작 흐름]
- 스프링 컨테이너는 빈 초기화 시점에 @Transactional 어노테이션이 붙은 빈에 대해 프록시를 생성함
- 클라이언트가 서비스를 호출하면, 실제로는 이 프록시 객체의 메서드가 호출됨
- 외부에서 메서드를 호출 할 떄 
     프록세 객체를 호출(트랜잭션 처리 로직) ->  프록시 객체는 실제 로직이 담긴 메서드를 호출
- 하지만 자기 자신을 내부에서 호출하는 메서드는 프록시를 거칠 수 없고 직접 호출된다.

즉, Self Invocation 이슈로 인해 트랜잭션이 정상적으로 동작하지 않았습니다.

<br>
<br>

---
이러한 문제를 해결하기 위해 다음과 같이  재시도 로직과 create...WithLock() 로직을 
`WaitingSaveService`로 따로 분리를 하고, `WaitingService`에 주입하여 
트랜잭션이 제대로 프록시 객체에 감싸져 동작 할 수 있도록 수정을 했어요!! 


해당 변경 사항은 
[fix: 내부 메서드 호출로 인해 트랜잭션이 동작하지 않던 문제 수정](https://github.com/next-step/spring-basic-roomescape-playground/pull/169/commits/c41865e7fa1f4d0fbde8e1591984eecc0d160030)
미션 제출 후에 다시 수정해서 커밋을 남겨 놓았어요

chxghee replied:
## 3️⃣ 비관락은 왜 예외를 던지는가?
> 그만큼 예외라함은 정상 상황 안에서 다른 분기라기 보단 비정상적인 상황이라고 생각하는 것이 적절하다고 봅니다.
....
비관락은 애초에 충돌이 발생할 가능성이 매우 크다는 가능성을 두고 예외를 던져버리는 기법입니다.
재시도를 할 수 있는 여러 방법은 존재하지만, 예외라는 큰 위험부담을 감수하고 이 예약의 순번을 부여해야하는지 잘 모르겠어요.


해당 로직에 예외를 캐치하고 재시도 로직을 넣었던 이유는 

- 처음 예약 대기를 등록할 때는 `WaitingOrderCounter`테이블에 튜플이 아직 없으니까
락의 대상이 되는 튜플을 생성해 주는 로직이 필요했었어요

- 그런데 이 상황에서도 많은 동시 요청이 온다면
같이 첫번째 튜플을 삽입하려 해 `DataIntegrityViolationException`의 발생하는 상황이 있을 것 같았어요

제가 질문을 잘 이해한진 모르겠지만 , 
이 재시도 로직이 처리하는 것은 락을 얻지 못했을 때 다시 얻기 위해가 아니라 
WaitingOrderCounter 테이블에 가장 첫 레코드를 삽입할때 충돌이 일어나게 되면,  이를 재시도 하기 위한 로직이었어요..!

그런데 

> 예외가 발생하면,
그대로 로직이 끝나고 예외 응답을 주기도 하고,
Transactinal 어노테이션은 예외를 catch 한다한들 rollback 플래그가 true로 바뀌기도 합니다.

에서 말씀해 주신 대로 위와 같은 예상하지 못한 상황이 발생 한다면,
예외를 잡고 재시도 하는 것 자체가 문제가 될 수가 있는 것 같아요

그렇다면 지금의 락을 거는 전략은 유지 한다고 했을 때,
재시도 로직을 없앨  고민을 해 보았는데, 
WaitingOrderCounter 테이블에 가장 첫 레코드를 삽입 하는 것을 지금의 방법처럼 `예약 대기` 요청 시 동시에 하는 것이 아니라,
`예약`이 성공 했을때 한다면 재시도 로직을 없앨 수 있을 것 같다 라는 생각이 듭니다!


---

그런데 이러한 질문들을 주신 이유가 근본적으로는 
`예약 대기`라는 요구사항을 해결하기 위해 락을 사용하는 것이 정말 좋은 방법인가?? 

에 대해 고민해 보라고 질문을 주신거라고 생각해요!

이것에 대해서는 두 가지 방법 모두 구현을 해보니 장단이 명확하게 체감이 되는거 같아요 ㅎㅎ 

[락을 쓸 때 장단점]
- 장점: 조회 성능을 올릴수 있다.
- 단점:  로직이 복잡하다., 저장, 삭제, 변경시 성능이 안 좋아 진다.

[락을 안 쓸 때 장단점] (조회시 대기 순위 계산)
- 장점: 로직이 간단하다., 순번 상태를 DB에 저장하지 않으니 사이드 이팩트를 줄일 수 있다.
- 단점: 조회 성능이 안좋다.


이렇게 정리 할 수 있을 것 같아요

어찌 되었든 정답이 있는 영역은 아닌 것 같고,
서비스의 성격에 따라 어떤 방식을 사용 할 지 골라야 하는 것 같아요

제가 결과적으로 락을 거는 전략을 선택한 이유는 
예약을 조회하는 횟수가 등록하는 횟수보다 훨씬 많을 것 같았기 떄문에 때문에 조회 성능을 올리고 싶어서 였습니다..!
(+ 동시성 문제에 대해 한번 깊게 고민하고, 적용해 보는 시간을 가져 보고 싶어서 라는 이유도 있긴 해요 😅)

---

경환님은 제 생각에 대해 어떻게 생각 하시나요??! 

락을 이용하실것 같나요?? 아님 제가 생각하지 못한 다른 방법이 있었을까요?!

dooboocookie replied:
## 🤔 Request Change

일단 여러 방법을 고민해보고 테스트를 짜보시면 좋겠어요.
진짜 500개 요청 밀어넣으면 500개 다 성공하는게 우린 목표였잖아요.

```
    @Test
    void 동시성_테스트_비관적락이_제대로_동작하는지_확인() throws InterruptedException {
        // given
        int threadCount = 500;
        ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
        CountDownLatch latch = new CountDownLatch(threadCount);
        AtomicInteger successCount = new AtomicInteger(0);
        AtomicInteger failureCount = new AtomicInteger(0);

        // when - 동시에 여러 대기 요청
        for (int i = 0; i < threadCount; i++) {
            final int index = i;
            executorService.submit(() -> {
                try {
                    Member member = memberRepository.save(new Member("user" + index, "user" + index + "@email.com", "1234", Role.USER));
                    WaitingCommand command = new WaitingCommand(date, member.getId(), theme.getId(), time.getId());

                    waitingSaveService.createWaiting(command, theme, time, member);
                    successCount.incrementAndGet();
                } catch (Exception e) {
                    failureCount.incrementAndGet();
                    System.out.println("Thread " + index + " failed: " + e.getMessage());
                } finally {
                    latch.countDown();
                }
            });
        }

        latch.await();
        executorService.shutdown();

        // then
        System.out.println("Success: " + successCount.get() + ", Failure: " + failureCount.get());

        // 실제 요청 수와 저장한 waiting 개수가 일치하는지 확인
        long waitingCount = waitingRepository.count();
        assertThat(waitingCount).isEqualTo(threadCount);
    }
```
<img width="252" height="94" alt="image" src="https://github.com/user-attachments/assets/2dbd2415-5b9f-4f53-b779-d894cc17359e" />

저장 로직에서 Thread.sleep으로 약간 지연시킨 후 위 테스트를 했을 때 결과는 다음과 같았어요.
[rollbackonly](https://techblog.woowahan.com/2606)에 라는 개념을 한번 살펴보면 좋겠습니다.

저도 이 리뷰에서 창희님이 동시성에 대해서 고려해본 경험 자체가 의미 있다고 생각합니다.
다만, 
- 락에 대상이 적합한가? 
- 다른 락 방법은? 
- 락 없이는? 
- 내가 짠 동시성 방어 로직이 잘 작동하나?

이런 내용을 더 고민해보셨으면 해서 드린 질문입니다.
제가 보기엔 이 로직은 잘 동작하지 않을 것으로 예상되거든요.

---

> [락을 쓸 때 장단점]
> 
>     장점: 조회 성능을 올릴수 있다.
>     단점: 로직이 복잡하다., 저장, 삭제, 변경시 성능이 안 좋아 진다.
> 
> [락을 안 쓸 때 장단점] (조회시 대기 순위 계산)
> 
>     장점: 로직이 간단하다., 순번 상태를 DB에 저장하지 않으니 사이드 이팩트를 줄일 수 있다.
>     단점: 조회 성능이 안좋다.

장단점에 대해서 대체적으로 동의됩니다.

하지만 더 중요한 문제는 이 자원은 읽기가 중요할까 쓰기가 중요할까 라고 생각했을 떄, 동시 쓰기에 대한 위기 감지를한것처럼 쓰기 성능 및 정확도가 매우매우 중요하겠죠.

일단 저라면, 만약 이런 경우라면 RDB를 쓰건 인메모리를 쓰건, 저장 당시 번호를 부여하지 않을것 같아요.
그냥 밀려들어오는 요청이 큐같은 FIFO 자료구조에 들어오는 것처럼 순서대로 적재해놓고 조회 시 계산해서 주거나 계산 내용을 따로 캐싱해둘것가타요.

chxghee replied:
우선 예시로 주신 테스트 코드를 통과할 수 있도록 기존의 비관적 락을 이용한 코드를 고쳐 보았어요!!

가장 큰 문제가 위에서 언급해 주신 [rollbackonly](https://techblog.woowahan.com/2606)에 라는 개념때문이었어요

블로그 글을 살펴 보니,
`@Transaction` 안에서 예외가 발생하면, 해당 예외를 catch를 한다 하더라도 
트랜잭션에 `rollback-only 마킹`을 해서 해당 트랜잭션은 롤백이 될수 밖에 없는 내부 동작이 이루어 지고 있었네요
(원래 트랜잭션의 재사용이 불가능한 상황이었습니다.)


[이전 코드의 문제]
- 떄문에 이전의 구조는 WaitingService의 `save` 안에서 WaitingSaveService의  `createWaiting`을 호출해

- 트랜잭션을 시작해서 , `락을 걸어 새로운 예약대기를 생성 및 저장 하는 로직` + `예외 발생시 재시도 로직`을 하도록 했는데요,

이때 `예외 발생시 재시도 로직`을 트랜잭션 안에서 실행을 하니, 
순위를 유지하는 카운터 삽입시 충돌이 발생하면 `rollback-only 마킹`이 되어 롤백이 일어나 재시도 로직이 제대로 동작하지 않게 되었던 상황이었습니다.
 
[해결한 방법]
- 트랜잭션 안의 재시도 로직을 `WaitingService` 로 옮겨, 트랜잭션 외부에서 예외를 잡고 재시도 로직을 수행 하도록 했습니다.

정리하자면  
WaitingService - `단순 조회 + 유효성 검사 + 재시도` 를 트랜잭션외부에서 하고, 
WaitingSaveService - 트랜잭션 내부에서는 락을 걸고 변경 내용을 수정하도록 해서 
 동시 요청 시에도 재시도 로직이 잘 동작 할 수 있게끔 수정했습니다!

dooboocookie commented on src/main/java/roomescape/theme/domain/ThemeRepository.java:
```diff
@@ -1,41 +1,28 @@
-                 .stream()
-                 .findFirst();
-    }
+    @Override
+    @Query("select t from Theme t where t.deleted = false")
+    List<Theme> findAll();
```
## 👀 Comment

deleted라는 boolean 변수를 사용해서 soft deleted를 구현하셨네요.
좋습니다. 몇가지 질문드립니다.

1. soft deleted를 사용하면서 느낀 장단점이 있을까요?
2. Theme에 Where 어노테이션 붙이는 것은 고려해봤는지?
3. findAllByDeletedIsFalse(사용 방법이 맞는지는 잘 모르겠음)이런 Spring Data의 Query method 방식은 고려해봤는지?

chxghee replied:
## 1. soft deleted를 사용하면서 느낀 장단점이 있을까요?

[장점]
1. 데이터 보존  
    - 데이터를 물리적으로 삭제하지 않기 때문에, 삭제된 데이터를 복구하거나 기록을 확인할 수 있다.

3. 삭제시 참조 관계 유지  
    - 외래 키로 해당 테이블을 참조하는 테이블이 있을 경우 삭제가 되어도, 참조하는 레코드에서 데이터를 유지할 수 있다.


[단점]
1. 조건절에 `deleted = false` 누락  
    - 모든 쿼리에 `deleted = false`를 붙여주어야 하기 때문에 실수로 뺴먹을 수 가 있다.

2. 성능    
    - DB에 레코드가 계속 쌓이게 되어 전체적으로 성능이 저하될 수 있다.

정도로 정리 할 수 있을 것 같아요~~

이번 미션에서는 장점 2번의 이유 때문에 사용했어요!
이번 미션에선 관리자는 Theme와 Time을 삭제할 수 있는데, 

이떄 관리자의 실수로 인해 잘못 삭제 하게 된다면, 모든 예약 데이터들이 제약 조건에 따라 같이 삭제가 되거나 데이터가 없어져 복구를 못할 수 있는 상황이 있을 수 있겠다 싶었어요\
(미션 원본 구조도 `soft delete`를 사용하고 있었는데, 아마 이러한 이유 때문에 사용한것 같아요.)



<br>

---
## 2. Theme에 Where 어노테이션 붙이는 것은 고려해봤는지?

이 방법은 언급을 해 주셔서 처음 알게 되었는데요,
아래처럼
```java
@Entity
@Where(clause = "deleted = false")
public class Theme {
    ...
}
```
엔티티 클래스레벨에 붙이는 옵션으로, 모든 쿼리의 조건절에 자동으로 `deleted = false`를 추가해 주는 군요!

근데 저는 이걸 알고 있었다 해도 쓰지는 않았을 것 같아요

[안쓰는 이유]
1. Repository에 명시적으로 써 있지 않아 실수 유발 가능
    - 쿼리 메서드 들에 `deleted = false`가 생략되어 있어 협업을 하는 상황이라면 많이 혼란을 줄수 있다.
    
2. Repository에서 유연하게 쿼리 생성가능  
    - 만약 딜리트된 것도 조회해야 하는 상황이 있다면 따로 번거로운 작업을 해아 한다.

이러한 이유 떄문에  Where 어노테이션 붙일  바에는 
차라리 Repository에 명시적으로 `deleted = false` 조건을 넣는것이 좋을것 같다 라고 생각을 해요!!
 


<br>

---

## 3. findAllByDeletedIsFalse(사용 방법이 맞는지는 잘 모르겠음)이런 Spring Data의 Query method 방식은 고려해봤는지?

이 방식은 고려해 보았는데, 

`findAll()`메서드는 기본으로 제공 되는데 
`soft deleted`가 적용된 상황을 모르고 `findAll()`메서드를 쓴다면 문제가 될 수도 있겠다 싶었어요.

때문에  위와 같은 사이드 이팩트를 사전에 막을 수 있게 findAll()  메서드를 오버라이딩 하는 방식으로 코드를 작성하게 되었습니닷!!

dooboocookie replied:
그래요. 소프트 딜리트에 대한 생각 정리 잘하신 것 같습니다.

말씀하신 장단점이 대체적으로 공감이 되기도하고 창희 스스로도 잘 정리하고 있는 것 같습니다.

다만,  `Repository에 명시적으로 써 있지 않아 실수 유발 가능`이렇게 말씀하셨는데, softdelete는 여러 이유에 의해서 실제 삭제를 하지 않는 것 뿐이지, 데이터가 비즈니스 적으로는 삭제되어야함은 마땅합니다.

그래서 `A쪽 실수 유발 방지`가 `B쪽 실수 유발`로 이어질 수 있겠죠.
누가 repository에 findByTime 같은 조회 로직을 정의하다가 deleted=false를 빼놓고 만들 수 있죠. 그럼 실제론 없는데이터가 조회될 수도 있겠죠.
A가 치명적일까 B가 더 치명적일까 어떤걸 막아야할까는 전적으로 창희님과 그 팀원들이 의논해서 결정하면 좋겠습니다.

chxghee commented:
추가로 하나 질문드릴게 생겼는데요, 

`@Transactional(readOnly = true)`를 단순 조회만 하는 서비스 메서드에 붙이는 것이 좋은가? 하는 고민을 하게 되었어요.

우선 `@Transactional(readOnly = true)`을 설정함으로써 얻을 수 있는 장단점은 

[장점]
1. Dirty Checking과 Flush() 생략 - 변경 감지 로직 자체를 생략 가능 → 성능을 최적화 할 수 있다.
2. 쿼리 계획 최적화 - DB에서 쿼리계획을 작성할 때 해당 쿼리가 읽기 전용인지 여부를 고려하여 쿼리계획을 최적화 할 수 있다.
3. Lazy Loading 가능 - 트랜잭션 내내 커넥션을 점유하기 때문에 객체를 사용하는 시점에서 쿼리를 날리는 Lazy Loading이 가능하다.


[단점]
1. 트랜잭션 내내 커넥션을 점유한다.


이 단점 떄문에 많은 요청이 몰리는 상황에서는 커넥션을 얻지 못해 대기해야 하는 요청이 생길 수 있을 것 같다는 생각이 들었어요

그래서 Lazy Loading을 하지 않는 단순 조회용 서비스 메서드에서는 `@Transactional(readOnly = true)` 설정을 하지 않는게 오히려 더 좋겠다 라는 생각이 들었어요

이에 대해서 어떻게 생각 하시나요?

chxghee committed:
fix: 내부 메서드 호출로 인해 트랜잭션이 동작하지 않던 문제 수정

WaitingSaveService, ReservationSaveService 로 별도 클래스로 분리하여
DB에 저장하는 로직에 대해 트랜잭션이 적용되도록 수정

dooboocookie reviewed (changes requested):
안녕하세요, 창희.
리뷰어 루카입니다.🐳
지난 번 회식 때 봤던 기억이 있네요.
제가 그 후로 밋업도 못가고, 리뷰도 매칭이 안돼서 처음 이야기 나눠보는 것 같네요.
일단 리뷰 들어가기에 전에 리뷰를 너무 늦게 드려 죄송해요.

## 🐳 리뷰 스타일

각 코멘트마다 3가지로 분류를 할 건데, 그 의미는 다음과 같습니다.

- Request Change
  - 버그 유발, 컨벤션 어김, 비즈니스 요구사항 어김, ...
  - 위와 같이 심각한 문제를 내포하고 있어, 어떤 관점으로 봐도 변경이 필요한 경우
- Comment
  - 반드시 변경할 필요는 없지만, 여러 관점으로 이야기 나눠볼만한 주제
- Approve
  - 동의되는 코드

## 🎯 리뷰 포인트
- 비관락을 통한 동시성 해결

---
이번 PR은 진짜 드릴 내용이 많지 않네요.
일단 이번 커밋들 변경 범위에서 코멘트 달 수 있는 내용이 많지 않아요.

그래서 비관락에 대한 내용만 일단 리뷰드릴테니,
이전에 리뷰들에서 했던 이야기지만 루카의 관점이 궁금한 내용이 있으면 말씀해주세요.

chxghee committed:
fix: 동시 예약 대기 요청 문제 해결
 발생한 예외를 잡고 재시도 하는 로직에서 트랜잭션의 rollback-only 마킹으로 인해
재시도 로직 실행시 트랜잭션을 재사용하지 못하던 문제를 해결하였습니다.
-> 재시도 로직을 Tx 밖으로 이동

chxghee committed:
test: 동시 요청 테스트 코드 작성

chxghee commented:
경환님
제출했던 코드에서 여러 쓰레드를 만들어 테스트를 했을 때, 제 로직이 잘 동작하지 않았던 문제를 찾아 해결했습니닷

아무래도 락에 대해 공부해 보고 적용해 본 것이 이번 미션이 처음이라 주신 질문들에 대해 선뜻 답하기가 쉽지 않더군요.

하지만 덕분에 동시성 문제를 다루는 여러 방식을 찾아보고,  
실제 적용해 보았을 때 오류 상황도 겪어 보면서 입문할 수 있었던것 같습니다.

수정한 제 코드에서 문제가 있거나 개선할 만한 점이 있다면 편하게 말씀해 주세요!

P.S. 주말인데도 시간 내어 리뷰해 주셔서 감사합니다 👍
